File: src/main/java/no/nav/veilarbaktivitet/aktivitetskort/service/TiltakMigreringCronService.java
package no.nav.veilarbaktivitet.aktivitetskort.service;

import lombok.RequiredArgsConstructor;
import net.javacrumbs.shedlock.spring.annotation.SchedulerLock;
import no.nav.common.featuretoggle.UnleashClient;
import no.nav.veilarbaktivitet.aktivitet.AktivitetDAO;
import no.nav.veilarbaktivitet.aktivitet.domain.AktivitetData;
import no.nav.veilarbaktivitet.aktivitet.domain.AktivitetTransaksjonsType;
import no.nav.veilarbaktivitet.util.DateUtils;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class TiltakMigreringCronService {

    private final UnleashClient unleashClient;

    private final TiltakMigreringDAO tiltakMigreringDAO;

    private final AktivitetDAO aktivitetDAO;

    public static final String TILTAK_HISTORISK_MIGRERING_CRON_TOGGLE = "veilarbaktivitet.tiltakmigrering.cron.disabled";

    @Scheduled(
            initialDelayString = "${app.env.scheduled.default.initialDelay}",
            fixedDelayString = "${app.env.scheduled.default.fixedDelay}"
    )
    @SchedulerLock(name = "historiske_tiltak_migrering_cron", lockAtMostFor = "PT20M")
    public void settTiltakOpprettetSomHistoriskTilHistorisk() {
        if (unleashClient.isEnabled(TILTAK_HISTORISK_MIGRERING_CRON_TOGGLE)) {
            return;
        }

        List<AktivitetData> aktiviteter = tiltakMigreringDAO.hentTiltakOpprettetSomHistoriskSomIkkeErHistorisk(500);

        aktiviteter
                .stream()
                .map(a -> a
                        .withTransaksjonsType(AktivitetTransaksjonsType.BLE_HISTORISK)
                        .withHistoriskDato(DateUtils.localDateTimeToDate(a.getEksternAktivitetData().getOppfolgingsperiodeSlutt())))
                .forEach(aktivitetDAO::oppdaterAktivitet);
    }

}

File: src/main/kotlin/no/nav/veilarbaktivitet/aktivitetskort/service/TiltakMigreringCronService.kt
package no.nav.veilarbaktivitet.aktivitetskort.service

import io.getunleash.Unleash
import no.nav.veilarbaktivitet.aktivitet.AktivitetDAO
import no.nav.veilarbaktivitet.aktivitet.domain.AktivitetData
import no.nav.veilarbaktivitet.aktivitet.domain.AktivitetTransaksjonsType
import no.nav.veilarbaktivitet.person.Innsender
import no.nav.veilarbaktivitet.util.DateUtils
import org.springframework.stereotype.Component
import java.util.*

@Component
open class TiltakMigreringCronService(
    val unleash: Unleash,
    val tiltakMigreringDAO: TiltakMigreringDAO,
    val aktivitetDAO: AktivitetDAO,
) {

    /* Trenger kun denne ved patching av historiske data fra arena */
//    @Scheduled(
//        initialDelayString = "\${app.env.scheduled.default.initialDelay}",
//        fixedDelayString = "\${app.env.scheduled.default.fixedDelay}",
//    )
//    @SchedulerLock(name = "historiske_tiltak_migrering_cron", lockAtMostFor = "PT20M")
    open fun settTiltakOpprettetSomHistoriskTilHistorisk() {
        if (unleash.isEnabled(TILTAK_HISTORISK_MIGRERING_CRON_TOGGLE)) return
        tiltakMigreringDAO.hentTiltakOpprettetSomHistoriskSomIkkeErHistorisk(500)
            .map { aktivitet -> aktivitet
                    .withTransaksjonsType(AktivitetTransaksjonsType.BLE_HISTORISK)
                    .withHistoriskDato(DateUtils.localDateTimeToDate(aktivitet.getEksternAktivitetData().oppfolgingsperiodeSlutt))
                    .withEndretDato(Date())
                    .withEndretAvType(Innsender.SYSTEM)
                    .withEndretAv("veilarbaktivitet")
            }
            .forEach { aktivitet: AktivitetData? -> aktivitetDAO.oppdaterAktivitet(aktivitet) }
    }

    companion object {
        const val TILTAK_HISTORISK_MIGRERING_CRON_TOGGLE = "veilarbaktivitet.tiltakmigrering.cron.disabled"
    }
}
